test/Validate.SSA.cpp

```diff
@@ -1172,6 +1173,67 @@ TEST_F(ValidateSSA, PhiUseDoesntDominateUseOfPhiOperandUsedBeforeDefinitionBad)
     MatchesRegex("ID .\\[inew\\] has not been defined"));
 }
 
+TEST_F(ValidateSSA, PhiUseMayComeFromNonDominatingBlockGood) {
+  string str = kHeader
+      + "OpName %if_true \"if_true\"\n"
+      + "OpName %exit \"exit\"\n"
+      + "OpName %copy \"copy\"\n"
+      + kBasicTypes +
+               R"(
+%func        = OpFunction %voidt None %vfunct
+%entry       = OpLabel
+               OpBranchConditional %false %if_true %exit
+
+%if_true     = OpLabel
+%copy        = OpCopyObject %boolt %false
+               OpBranch %exit
+
+; The use of %copy here is ok, even though it was defined
+; in a block that does not dominate %exit.  That's the point
+; of an OpPhi.
+%exit        = OpLabel
+%value       = OpPhi %boolt %false %entry %copy %if_true
+               OpReturn
+               OpFunctionEnd
+)";
+
+  CompileSuccessfully(str);
+  ASSERT_EQ(SPV_SUCCESS, ValidateInstructions()) << getDiagnosticString();
+}
+
+TEST_F(ValidateSSA, DISABLED_PhiVariableDefMustComeFromBlockDominatingThePredecessorBad) {
```

Comment:
It's a requirement for SSA that it work that way. 
The SPIR-V spec says the module must be in SSA form but doesn't spell out all the rules for SSA.
The best I can point to is the section on SSA form: https://www.khronos.org/registry/spir-v/specs/1.1/SPIRV.html#_a_id_ssa_a_ssa_form  especially at the end where it mentions the use of a Phi.

Also, in the universal validation rules under the SSA bullet it says "Uses in a phi-function in a loop may consume definitions in the loop that don’t dominate the use."  Which is true (and vague) but also leaves out the forward-only but divergent then reconvergent control flow case like in this example (diamond structure).